packages/assets-controllers/src/TokenBalancesController.ts

```diff
@@ -8,134 +12,223 @@ import type { AssetsContractController } from './AssetsContractController';
-      this.updateBalances();
+  constructor({
+    interval = DEFAULT_INTERVAL,
+    tokens = [],
+    disabled = false,
+    onTokensStateChange,
+    getSelectedAddress,
+    getERC20BalanceOf,
+    messenger,
+    state = {},
+  }: TokenBalancesControllerOptions) {
+    super({
+      name: controllerName,
+      metadata,
+      messenger,
+      state: {
+        ...getDefaultState(),
+        ...state,
+      },
     });
-    this.getSelectedAddress = getSelectedAddress;
-    this.getERC20BalanceOf = getERC20BalanceOf;
+
+    this.#disabled = disabled;
+    this.#interval = interval;
+    this.#tokens = tokens;
+
+    onTokensStateChange(this.#tokensStateChangeListener.bind(this));
+
+    this.#getSelectedAddress = getSelectedAddress;
+    this.#getERC20BalanceOf = getERC20BalanceOf;
+
     this.poll();
   }
 
+  /*
+   * Tokens state changes listener.
+   */
+  #tokensStateChangeListener({ tokens, detectedTokens }: TokensState) {
+    this.#tokens = [...tokens, ...detectedTokens];
+    this.updateBalances();
+  }
+
+  /**
+   * Allows controller to update tracked tokens contract balances.
+   */
+  enable() {
+    this.#disabled = false;
+  }
+
+  /**
+   * Blocks controller from updating tracked tokens contract balances.
+   */
+  disable() {
+    this.#disabled = true;
+  }
+
+  /*
+   * Lists all tracked tokens.
+   */
+  getTokens() {
+    return this.#tokens;
+  }
+
   /**
    * Starts a new polling interval.
    *
    * @param interval - Polling interval used to fetch new token balances.
    */
   async poll(interval?: number): Promise<void> {
-    interval && this.configure({ interval }, false, false);
-    this.handle && clearTimeout(this.handle);
+    if (interval) {
+      this.#interval = interval;
+    }
+
+    if (this.#handle) {
+      clearTimeout(this.#handle);
+    }
+
     await safelyExecute(() => this.updateBalances());
-    this.handle = setTimeout(() => {
-      this.poll(this.config.interval);
-    }, this.config.interval);
+
+    this.#handle = setTimeout(() => {
+      this.poll(this.#interval);
+    }, this.#interval);
   }
 
   /**
    * Updates balances for all tokens.
    */
   async updateBalances() {
-    if (this.disabled) {
+    if (this.#disabled) {
       return;
     }
-    const { tokens } = this.config;
-    const newContractBalances: { [address: string]: BN } = {};
-    for (const token of tokens) {
+
+    const newContractBalances: ContractBalances = {};
+    for (const token of this.#tokens) {
       const { address } = token;
       try {
-        newContractBalances[address] = await this.getERC20BalanceOf(
-          address,
-          this.getSelectedAddress(),
-        );
+        newContractBalances[address] = (
+          await this.#getERC20BalanceOf(address, this.#getSelectedAddress())
+        ).toString(16);
         token.balanceError = null;
       } catch (error) {
-        newContractBalances[address] = new BN(0);
+        newContractBalances[address] = new BN(0).toString(16);
```

Comment:
new BN(0).toString(16) return '0' ?

Reply:
Ah. You're right. That raises a point — if we use `toString(16)`, the resulting string won't have a `0x` prefix. This could be potentially confusing. The reason for this is that the hex representation of some numbers are ambiguous. For instance, 20 in hex is not 20 in decimal. Therefore, someone reading these values directly in state may be confused. But if we prefix every hex value with `0x` then it is clear what it is.

The `controller-utils` package has a `toHex` function that takes a BN object and returns a `0x`-prefixed string. So in this case we could say:

``` suggestion
        newContractBalances[address] = toHex(0);
```